### ISSUE TITLE
x509 SAN UPN other name is not handled in JDK 21

### ISSUE THREAD

**Issue by mabartos** [keycloak/keycloak#29968]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

When dealing with the [support for JDK 21](https://github.com/keycloak/keycloak/pull/28518), I've encountered an issue with the x509 Subject Alternative Name (SAN) User Principal Name (`otherName`) in certificates. 

It seems the attribute is not properly handled when using JDK 21. It works properly with JDK 17.

To check the SAN props of certificate with `openssl x509 -noout -text -in test-user-san@localhost.cert.pem`:

```
X509v3 Subject Alternative Name: 
                email:test-user-altmail@localhost, IP Address:192.168.7.1, DNS:www.example-test.com, URI:http://www.example-test.com, othername: 1.2.3.4::my_test_user, othername: UPN::test_upn_name@localhost
```

It was caught by test `X509BrowserLoginSubjectAltNameTest#loginAsUserFromCertSANUpn` and log with the failed test:

```
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,578 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] [X509BrowserLoginSubjectAltNameTest] loginAsUserFromCertSANUpn() STARTED
X509BrowserLoginSubjectAltNameTest ++ May 29, 2024 8:50:02 AM org.jboss.arquillian.drone.webdriver.factory.HtmlUnitDriverFactory setClientOptions
X509BrowserLoginSubjectAltNameTest ++ INFO: Setting HtmlDriver web client options: {htmlunitCssEnabled=false, htmlunitHistoryPageCacheLimit=1}
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,588 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Configuring driver settings. implicitWait=2000, pageLoadTimeout=10000
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,588 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Check HtmlUnit driver TLS settings
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,588 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Keystore '/home/runner/work/keycloak/keycloak/testsuite/integration-arquillian/tests/base/target/containers/auth-server-quarkus/conf/certs/clients/test-user-san-cert-test-user-key@localhost.p12', password 'password', type 'pkcs12'
X509BrowserLoginSubjectAltNameTest ++ May 29, 2024 8:50:02 AM org.jboss.arquillian.drone.webdriver.factory.HtmlUnitDriverFactory setClientOptions
X509BrowserLoginSubjectAltNameTest ++ INFO: Setting HtmlDriver web client options: {htmlunitCssEnabled=false, htmlunitHistoryPageCacheLimit=1}
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,614 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Configuring driver settings. implicitWait=2000, pageLoadTimeout=10000
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,615 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Check HtmlUnit driver TLS settings
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,615 INFO  [org.keycloak.testsuite.drone.KeycloakDronePostSetup] Keystore '/home/runner/work/keycloak/keycloak/testsuite/integration-arquillian/tests/base/target/containers/auth-server-quarkus/conf/certs/clients/test-user-san-cert-test-user-key@localhost.p12', password 'password', type 'pkcs12'
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,656 DEBUG [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] loading test realms
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,656 DEBUG [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] Adding test realm for import from testrealm.json
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,657 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] importing test realms
X509BrowserLoginSubjectAltNameTest ++ 08:50:02,657 DEBUG [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] --importing realm: test
X509BrowserLoginSubjectAltNameTest ++ 2024-05-29 08:50:04,299 ERROR [org.keycloak.services] (executor-thread-2) [X509ClientCertificateAuthenticator:authenticate] Exception: class java.lang.String cannot be cast to class [B (java.lang.String and [B are in module java.base of loader 'bootstrap')
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,310 INFO  [org.keycloak.testsuite.util.URLUtils] starting navigation
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,310 INFO  [org.keycloak.testsuite.util.URLUtils] current URL:  https://localhost:8543/auth/realms/test/protocol/openid-connect/auth?response_type=code&client_id=test-app&redirect_uri=https%3A%2F%2Flocalhost%3A8543%2Fauth%2Frealms%2Fmaster%2Fapp%2Fauth&state=693b74d2-2a3a-4a53-bde5-a0e0acfa1bed&scope=openid
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,310 INFO  [org.keycloak.testsuite.util.URLUtils] navigating to https://localhost:8543/auth/realms/test/testing/blank
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.util.URLUtils] new current URL:  https://localhost:8543/auth/realms/test/testing/blank
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.util.URLUtils] navigation complete
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] deleting cookies in 'test' realm
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] removing test realms after test method
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,316 INFO  [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] removing realm: test
X509BrowserLoginSubjectAltNameTest ++ 08:50:06,514 ERROR [org.keycloak.testsuite.x509.X509BrowserLoginSubjectAltNameTest] [X509BrowserLoginSubjectAltNameTest] loginAsUserFromCertSANUpn() FAILED
...
...
Error: OR]   Run 1: X509BrowserLoginSubjectAltNameTest.loginAsUserFromCertSANUpn:57->AbstractX509AuthenticationTest.x509BrowserLogin:571 » NoSuchElement Unable to locate element with ID: 'certificate_subjectDN'
-- For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
-- Build info: version: '4.21.0', revision: '79ed462ef4'
-- System info: os.name: 'Linux', os.arch: 'amd64', os.version: '6.5.0-1021-azure', java.version: '21.0.3'
-- Driver info: driver.version: HtmlUnitDriver
```

https://github.com/keycloak/keycloak/actions/runs/9271211106/job/25541538448?pr=28518

It seems there's certain improper type casting based on the log (not shown on local machine):
```
X509BrowserLoginSubjectAltNameTest ++ 2024-05-29 08:50:04,299 ERROR [org.keycloak.services] (executor-thread-2) [X509ClientCertificateAuthenticator:authenticate] Exception: class java.lang.String cannot be cast to class [B (java.lang.String and [B are in module java.base of loader 'bootstrap')
```  

Even though there were recent changes(https://github.com/keycloak/keycloak/commit/76a6733f0a0edf25f3e75724b0cf708f0284e19f#diff-9e112d961710fdb5fb2821f23a9c732eab5d5fc4e4f6c89d9157d3eb46597e43R47) in the test class, it should reflect the situation same as before as the certificates are identical (checked by the openssl tool).

In normal situations, the page should redirect to the `login-x509-info.ftl`, but with the JDK 21 and the specific test case, the page is redirected to the login screen and requires authentication.

@mposolda @rmartinc Could you please check it? 

### Version

main

### Regression

- [ ] The issue is a regression

### Expected behavior

Test pass

### Actual behavior

Test does not pass

### How to Reproduce?

1. Checkout to https://github.com/keycloak/keycloak/pull/28518 branch in order to properly execute these tests
2. Change Java version to 21 (_mvn needs to reflect it as well_)
3. `mvn clean install -Dtest=X509BrowserLoginSubjectAltNameTest#loginAsUserFromCertSANUpn -f testsuite/integration-arquillian/tests/base/pom.xml`

**Comment by rmartinc** [comment]
I have taken a look and you just need the following:

```diff
diff --git a/crypto/default/src/main/java/org/keycloak/crypto/def/BCUserIdentityExtractorProvider.java b/crypto/default/src/main/java/org/keycloak/crypto/def/BCUserIdentityExtractorProvider.java
index 13302ef4d9..a016e72a9d 100644
--- a/crypto/default/src/main/java/org/keycloak/crypto/def/BCUserIdentityExtractorProvider.java
+++ b/crypto/default/src/main/java/org/keycloak/crypto/def/BCUserIdentityExtractorProvider.java
@@ -149,6 +149,7 @@ public class BCUserIdentityExtractorProvider  extends UserIdentityExtractorProvi
                                 return obj;
                             }
 
+                            if (obj instanceof byte[]) {
                                 byte[] otherNameBytes = (byte[]) obj;
 
                                 try {
@@ -179,6 +180,7 @@ public class BCUserIdentityExtractorProvider  extends UserIdentityExtractorProvi
                                     logger.error("Failed to parse subjectAltName", e);
                                 }
                             }
+                        }
 
                     }
                 }
```

The javadoc says that it can return a String or a byte[]. So just checking it's a byte[] is enough for me to make the test pass OK.

**Comment by mposolda** [comment]
@mabartos Sending PR here to your branch https://github.com/mabartos/keycloak/pull/904.

As pointed by @rmartinc, there is the change in the behaviour. The Java 21 tries to be more "friendly" and hence it adds some additional items to the list containing String with parsed type-oid and possibly also the name. There is some small change in the javadoc of this method https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/security/cert/X509Certificate.html#getSubjectAlternativeNames() . Especially in Java 21, there are those additional lines adding in the javadoc:
```
A third entry may also be present in the list containing the type-id of the otherName in string form, and a fourth entry containing its 
value as either a string (if the value is a valid supported character string) or a byte array containing the ASN.1 DER encoded form 
of the value without the context-specific constructed tag with number 0. 
```

I've tested also in FIPS environment and figured it is not reliable to rely on those Strings being always present in Java 21. So just did what @rmartinc suggested and still doing the parsing of "type-oid" and subjectName ourselves from the byte array as we're doing for Java 17.

Updated both FIPS and non-fips version of this class.
